src/main/java/subproblem/Tree.java

```diff
@@ -30,60 +29,120 @@ public void addNode(Node node) {
     }
 
     // TODO: Implement (KP)
-    public List<Node> findShortestPath() {
-
+    public List<Node> findShortestPath(Node root) {
+        LinkedList<Node> queue = new LinkedList<>();
         List<Node> shortestPath = new LinkedList<>();
-        return shortestPath;
-    }
 
-    // TODO: Implement (KP)
-    public void breadthFirstSearch(Node root) {
-        LinkedList<Node> queue = new LinkedList<>();
+        double globalBestCost = Double.POSITIVE_INFINITY;
 
         queue.add(root);
         root.setVisited(true);
 
-        while(queue.size() != 0) {
-            Node s = queue.removeFirst();
-            Set<Node> setOfChildren = s.getChildren();
+        // This can be done when creating the root node
+        List<Node> rootPath = new ArrayList<>();
+        rootPath.add(root);
+        root.setBestPath(rootPath);
+        root.setBestCost(0.0);
+
+        while (queue.size() != 0) {
+            Node currentNode = queue.removeFirst();
+            Set<Node> setOfChildren = currentNode.getChildren(); // Retrieved in order right to left in tree
+
+            if (currentNode != root) {
+                currentNode.setBestCost(Double.POSITIVE_INFINITY);
+                for (Node parent : currentNode.getParents()) {
+                    double currentCost = parent.getBestCost() + parent.getCostOfChild(currentNode);
+                    if (currentCost < currentNode.getBestCost()) {
+                        currentNode.setBestCost(currentCost);
+                        List<Node> path = deepCopy(parent.getBestPath());  // TODO: Research best way to copy
+                        path.add(currentNode);
+                        currentNode.setBestPath(path);
+                    }
+                }
+            }
+
+            if(setOfChildren.isEmpty()) {
+                if(currentNode.getBestCost() < globalBestCost) {
+                    globalBestCost = currentNode.getBestCost();
+                    List<Node> globalBestPath = deepCopy(currentNode.getBestPath());
+                    shortestPath = globalBestPath;
+                }
```

Comment:
Lag en egen funksjon for dette, typ updateGlobalBest
Når du lager funksjonen, sjekk at den faktisk oppdaterer globalBestCost og shortestPath (som gis inn som argumenter til funksjonen) ved å printe verdiene av de variablene før og etter funksjonskallet.